service/popm/popm.go

```diff
@@ -785,3 +841,75 @@ func (m *Miner) Run(pctx context.Context) error {
 
 	return err
 }
+
+func (m *Miner) AddL2Keystone(val hemi.L2Keystone) {
+	serialized := hemi.L2KeystoneAbbreviate(val).Serialize()
+	key := hex.EncodeToString(serialized[:])
+
+	toInsert := L2KeystoneProcessingContainer{
+		l2Keystone:         val,
+		requiresProcessing: true,
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	// keystone already exists, no-op
+	if _, ok := m.l2Keystones[key]; ok {
+		return
+	}
+
+	if len(m.l2Keystones) < l2KeystonesMaxSize {
+		m.l2Keystones[key] = toInsert
+		return
+	}
+
+	var smallestL2BlockNumber uint32
+	var smallestKey string
+
+	for k, v := range m.l2Keystones {
+		if smallestL2BlockNumber == 0 || v.l2Keystone.L2BlockNumber < smallestL2BlockNumber {
+			smallestL2BlockNumber = v.l2Keystone.L2BlockNumber
+			smallestKey = k
+		}
+	}
+
+	// do not insert an L2Keystone that is older than all of the ones already
+	// added
+	if val.L2BlockNumber < smallestL2BlockNumber {
+		return
+	}
+
+	delete(m.l2Keystones, smallestKey)
+
+	m.l2Keystones[key] = toInsert
+}
+
+// l2KeystonesForProcessing creates copies of the l2 keystones, set them to
+// "processing", then returns the copies with the newest first
+func (m *Miner) l2KeystonesForProcessing() []hemi.L2Keystone {
+	m.mtx.Lock()
+
+	copies := []hemi.L2Keystone{}
```

Comment:
This should be allocated with make outside of the mutex.

Reply:
sure 👍